homeassistant/components/dlna_dmr/media_player.py

```diff
@@ -24,298 +27,499 @@
-):
-    """Register notify view."""
-    hass_data = hass.data[DLNA_DMR_DATA]
-    if "event_handler" in hass_data:
-        return hass_data["event_handler"]
-
-    # start event handler
-    server = AiohttpNotifyServer(
-        requester,
-        listen_port=server_port,
-        listen_host=server_host,
-        callback_url=callback_url_override,
+    entry: config_entries.ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the DlnaDmrEntity from a config entry."""
+    del hass  # Unused
+    _LOGGER.debug("media_player.async_setup_entry %s (%s)", entry.entry_id, entry.title)
+
+    # Create our own device-wrapping entity
+    entity = DlnaDmrEntity(
+        udn=entry.data[CONF_DEVICE_ID],
+        device_type=entry.data[CONF_TYPE],
+        name=entry.title,
+        event_port=entry.options.get(CONF_LISTEN_PORT) or 0,
+        event_callback_url=entry.options.get(CONF_CALLBACK_URL_OVERRIDE),
+        poll_availability=entry.options.get(CONF_POLL_AVAILABILITY, False),
+        location=entry.data[CONF_URL],
     )
-    await server.start_server()
-    _LOGGER.info("UPNP/DLNA event handler listening, url: %s", server.callback_url)
-    hass_data["notify_server"] = server
-    hass_data["event_handler"] = server.event_handler
-
-    # register for graceful shutdown
-    async def async_stop_server(event):
-        """Stop server."""
-        _LOGGER.debug("Stopping UPNP/DLNA event handler")
-        await server.stop_server()
-
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_stop_server)
-
-    return hass_data["event_handler"]
-
-
-async def async_setup_platform(
-    hass: HomeAssistant, config, async_add_entities, discovery_info=None
-):
-    """Set up DLNA DMR platform."""
-    if config.get(CONF_URL) is not None:
-        url = config[CONF_URL]
-        name = config.get(CONF_NAME)
-    elif discovery_info is not None:
-        url = discovery_info["ssdp_description"]
-        name = discovery_info.get("name")
-
-    if DLNA_DMR_DATA not in hass.data:
-        hass.data[DLNA_DMR_DATA] = {}
-
-    if "lock" not in hass.data[DLNA_DMR_DATA]:
-        hass.data[DLNA_DMR_DATA]["lock"] = asyncio.Lock()
-
-    # build upnp/aiohttp requester
-    session = async_get_clientsession(hass)
-    requester = AiohttpSessionRequester(session, True)
-
-    # ensure event handler has been started
-    async with hass.data[DLNA_DMR_DATA]["lock"]:
-        server_host = config.get(CONF_LISTEN_IP)
-        if server_host is None:
-            server_host = await async_get_source_ip(hass, PUBLIC_TARGET_IP)
-        server_port = config.get(CONF_LISTEN_PORT, DEFAULT_LISTEN_PORT)
-        callback_url_override = config.get(CONF_CALLBACK_URL_OVERRIDE)
-        event_handler = await async_start_event_handler(
-            hass, server_host, server_port, requester, callback_url_override
+
+    entry.async_on_unload(
+        entry.add_update_listener(entity.async_config_update_listener)
```

Comment:
All entity state update callbacks must be registered in `async_added_to_hass` and unregistered on entity remove.